Tools/clinic/clinic.py

```diff
@@ -910,54 +940,98 @@ def deprecate_positional_use(
-            )
-
+                amount = f"more than {first_pos} "
+                if first_pos == 1:
+                    what = "argument"
+            message = f"Parameters {pstr} will become keyword-only parameters"
+        message = (
+            f"Passing {amount}positional {what} to {func.fulldisplayname}() is deprecated. "
+            f"{message} in Python {major}.{minor}."
+        )
         # Append deprecation warning to docstring.
-        lines = textwrap.wrap(f"Note: {depr_message}")
-        docstring = "\n".join(lines)
+        docstring = textwrap.fill(f"Note: {message}")
         func.docstring += f"\n\n{docstring}\n"
+        # Format and return the code block.
+        code = self.DEPRECATION_WARNING_PROTOTYPE.format(
+            condition=condition,
+            errcheck="",
+            message=wrapped_c_string_literal(message, width=64,
+                                             subsequent_indent=20),
+        )
+        return normalize_snippet(code, indent=4)
+
+    def deprecate_keyword_use(
+            self,
+            func: Function,
+            params: dict[int, Parameter],
+    ) -> str:
+        assert len(params) > 0
+        names = [repr(p.name) for p in params.values()]
+        first_pos, first_param = next(iter(params.items()))
+        last_pos, last_param = next(reversed(params.items()))
+
+        # Pretty-print list of names.
+        pstr = pprint_words(names)
 
+        # For now, assume there's only one deprecation level.
+        assert first_param.deprecated_keyword == last_param.deprecated_keyword
+        thenceforth = first_param.deprecated_keyword
+        assert thenceforth is not None
+        major, minor = thenceforth
+
+        # Format the deprecation message.
+        containscheck = ""
+        conditions = []
+        for i, p in params.items():
+            if p.is_optional():
+                #conditions.append(f"nargs < {i+1} && {argname_fmt % i}")
+                if func.kind.new_or_init:
+                    conditions.append(f"nargs < {i+1} && kwargs && PyDict_Contains(kwargs, &_Py_ID({p.name}))")
+                    containscheck = "PyDict_Contains"
+                else:
+                    conditions.append(f"nargs < {i+1} && kwnames && PySequence_Contains(kwnames, &_Py_ID({p.name}))")
+                    containscheck = "PySequence_Contains"
+            else:
+                conditions = [f"nargs < {i+1}"]
+        condition = ") || (".join(conditions)
+        if len(conditions) > 1:
+            condition = f"({condition})"
+        if len(params) == 1:
+            what1 = "argument"
+            what2 = "parameter"
+        else:
+            what1 = "arguments"
+            what2 = "parameters"
+        message = (
+            f"Passing keyword {what1} {pstr} to {func.fulldisplayname}() is deprecated. "
+            f"Corresponding {what2} will become positional-only in Python {major}.{minor}."
+        )
```

Comment:
We should also produce a deprecation note for the docstring, like we do in `deprecate_positional_use()`.